File: Week2/공통/덧칠하기/suhyun1.py
# end_painted - s + 1 = m
# section에서 원소들을 차례로 꺼내, 꺼낸 원소부터 색칠할 수 있는 마지막 구간까지 칠한다고 가정한다.
def solution(n, m, section):
    answer, end_painted = 0, 0
    for s in section:
        # 이미 칠한 구간은 포함시키지 않고, 칠하지 않았다면, 다시 꺼낸 원소부터 칠할 수 있는 구간까지 칠한다.
        if s > end_painted:
            end_painted = m + s - 1
            answer += 1
    return answer

File: Week4/공통/디펜스게임/suhyun1.java
package Week4.공통.디펜스게임;

import java.util.PriorityQueue;

public class suhyun1 {
    public int solution(int n, int k, int[] enemy) {
        int answer = 0;

        PriorityQueue<Integer> pq = new PriorityQueue<>();

        for(int i=0; i<enemy.length; i++){
            if(k > 0){
                pq.offer(enemy[i]);
                k--;
            } else {
                //현재 적
                int num = enemy[i];

                //무적권을 사용했다고 가정한 병사의 수와 비교했을떄,
                //현재 적이 더 많다면,
                //현재 적에게 무적권을 사용하도록 수정
                if(num > pq.peek()){
                    num = pq.poll();
                    pq.offer(enemy[i]);
                }

                //이때의 num은
                //현재 적이거나
                //무적권을 사용했다고 가정한 수 중 가장 작은 원소
                //둘중 하나일 수 있다.
                if(n >= num){
                    n -= num;
                } else break;
            }
            answer++;
        }

        return answer;
    }
}
